hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -3131,3 +3156,142 @@ func reconcileMachineApproverDeployment(deployment *appsv1.Deployment, hc *hyper
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL, nil)
+		if err != nil {
+			return fmt.Errorf("failed to construct request for %s: %w", requestURL, err)
+		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			return fmt.Errorf("failed to do request for %s: %w", requestURL, err)
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("request to %s returned with unexpected status code %d", requestURL, resp.StatusCode)
+		}
+
+		// In theory we should be able to pass resp.Body
+		// as the Body argument to PutObject as that takes
+		// an io.Reader. In practice doing so results in
+		// `NotImplemented: A header you provided implies functionality that is not implemented`
+		// which apparently happens when there is no body:
+		// https://github.com/aws/aws-sdk-js/issues/15#issuecomment-11666580
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("faild to read body: %w", err)
+		}
+
+		_, err = r.S3Client.PutObject(&s3.PutObjectInput{
+			ACL:    aws.String("public-read"),
+			Body:   aws.ReadSeekCloser(bytes.NewReader(body)),
+			Bucket: aws.String(r.AWSOIDCBucketName),
+			Key:    aws.String(hcluster.Spec.InfraID + path),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to upload %s to the %s s3 bucket: %w", path, r.AWSOIDCBucketName, err)
+		}
+	}
+
+	hcluster.Finalizers = append(hcluster.Finalizers, oidcDocumentsFinalizer)
+	if err := r.Client.Update(ctx, hcluster); err != nil {
+		return fmt.Errorf("failed to update the hosted cluster after adding the %s finalizer: %w", oidcDocumentsFinalizer, err)
+	}
+
+	r.Log.Info("Successfully uploaded the OIDC documents to the S3 bucket")
+
+	return nil
+}
+
+func (r *HostedClusterReconciler) cleanupOIDCBucketData(ctx context.Context, hcluster *hyperv1.HostedCluster) error {
+	if !sets.NewString(hcluster.Finalizers...).Has(oidcDocumentsFinalizer) {
+		return nil
+	}
+
+	if r.AWSOIDCBucketName == "" || r.S3Client == nil {
+		return fmt.Errorf("hypershift wasn't configured with AWS credentials and a bucket, can not clean up OIDC documents from bucket. Please either set those up or clean up manually and then remove the %s finalizer from the hosted cluster", oidcDocumentsFinalizer)
+	}
+
+	var objectsToDelete []*s3.ObjectIdentifier
+	for _, path := range oidcPublicDocumentPaths() {
+		objectsToDelete = append(objectsToDelete, &s3.ObjectIdentifier{
+			Key: aws.String(hcluster.Spec.InfraID + path),
+		})
+	}
+
+	if _, err := r.S3Client.DeleteObjects(&s3.DeleteObjectsInput{
+		Bucket: aws.String(r.AWSOIDCBucketName),
+		Delete: &s3.Delete{Objects: objectsToDelete},
+	}); err != nil {
+		return fmt.Errorf("failed to delete OIDC objects from %s S3 bucket: %w", r.AWSOIDCBucketName, err)
+	}
+
+	hcluster.Finalizers = sets.NewString(hcluster.Finalizers...).Delete(oidcDocumentsFinalizer).UnsortedList()
```

Comment:
nit: could align with existing pattern in other controllers using controller runtime lib
```
		if controllerutil.ContainsFinalizer(hcluster, finalizer) {
			controllerutil.RemoveFinalizer(hcluster, finalizer)
```

Reply:
Done